Sources/PointFree/Auth.swift

```diff
@@ -82,94 +85,91 @@ public func fetchUser<A>(_ conn: Conn<StatusLineOpen, T2<Models.User.ID, A>>)
+      email: email,
+      now: { now }
+    )
+    await fireAndForget {
+      try await sendEmail(
+        to: [email],
+        subject: "Point-Free Registration",
+        content: inj2(registrationEmailView(env.gitHubUser))
+      )
     }
+    return user
+  } catch let PostgresError.server(error) where error.fields[.constraintName] == "users_email_key" {
+    throw GitHubUser.AlreadyRegistered(email: email)
+  }
 }
 
-/// Exchanges a github code for an access token and loads the user's data.
+/// Exchanges a GitHub code for an access token and loads the user's data.
 private func gitHubAuthTokenMiddleware(
   _ conn: Conn<StatusLineOpen, Tuple2<GitHub.AccessToken, String?>>
-)
-  -> IO<Conn<ResponseEnded, Data>>
-{
+) async -> Conn<ResponseEnded, Data> {
+  @Dependency(\.fireAndForget) var fireAndForget
   @Dependency(\.gitHub) var gitHub
   @Dependency(\.siteRouter) var siteRouter
 
   let (token, redirect) = lower(conn.data)
 
-  return EitherIO { try await gitHub.fetchUser(token) }
-    .map { user in GitHubUserEnvelope(accessToken: token, gitHubUser: user) }
-    .flatMap(fetchOrRegisterUser(env:))
-    .flatMap { user in
-      refreshStripeSubscription(for: user)
-        .map(const(user))
+  do {
+    let gitHubUser = try await gitHub.fetchUser(token)
+    let env = GitHubUserEnvelope(accessToken: token, gitHubUser: gitHubUser)
+    let user = try await fetchOrRegisterUser(env: env)
+    try await refreshStripeSubscription(for: user)
+    return conn.redirect(to: redirect ?? siteRouter.path(for: .home)) {
+      $0.writeSessionCookie { $0.user = .standard(user.id) }
     }
-    .withExcept(notifyError(subject: "GitHub Auth Failed"))
-    .run
-    .flatMap(
-      either(
-        const(
-          conn
-            |> PointFree.redirect(
-              to: .home,
-              headersMiddleware: flash(
-                .error,
-                "We were not able to log you in with GitHub. Please try again."
-              )
-            )
-        )
-      ) { user in
-        conn
-          |> HttpPipeline.redirect(
-            to: redirect ?? siteRouter.path(for: .home),
-            headersMiddleware: writeSessionCookieMiddleware { $0.user = .standard(user.id) }
-          )
-      }
-    )
+  } catch let error as GitHubUser.AlreadyRegistered {
+    return conn.redirect(to: .home) {
+      $0.flash(
+        .error, """
+        The primary email address associated with your GitHub account, \(error.email.rawValue), is \
+        already registered with Point-Free under a different \
+        [GitHub account](https://github.com/settings) account.
+
+        Log into the GitHub account associated with your Point-Free account before trying again, \
+        or contact <support@pointfree.co>.
+        """
+      )
+    }
+  } catch {
+    await fireAndForget {
+      try await sendEmail(
+        to: adminEmails,
+        subject: "GitHub Auth Failed",
+        content: inj1(String(customDumping: error))
+      )
+    }
```

Comment:
We'll no longer get admin emails for the duplicate email case, so it'll be on the user to reach out to us based off the flash message.